modules/ngx_http_proxy_connect_module/ngx_http_proxy_connect_module.c

```diff
@@ -0,0 +1,2129 @@
+            return;
+        }
+
+        ngx_log_error(NGX_LOG_INFO, ev->log, ev->kq_errno,
+                      "kevent() reported that client prematurely closed "
+                      "connection");
+
+        if (u->peer.connection == NULL) {
+            ngx_http_proxy_connect_finalize_request(r, u,
+                                               NGX_HTTP_CLIENT_CLOSED_REQUEST);
+        }
+
+        return;
+    }
+
+#endif
+
+    n = recv(c->fd, buf, 1, MSG_PEEK);
+
+    err = ngx_socket_errno;
+
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, ev->log, err,
+                   "http proxy_connect upstream recv(): %d", n);
+
+    if (ev->write && (n >= 0 || err == NGX_EAGAIN)) {
+        return;
+    }
+
+    if ((ngx_event_flags & NGX_USE_LEVEL_EVENT) && ev->active) {
+
+        event = ev->write ? NGX_WRITE_EVENT : NGX_READ_EVENT;
+
+        if (ngx_del_event(ev, event, 0) != NGX_OK) {
+            ngx_http_proxy_connect_finalize_request(r, u,
+                                               NGX_HTTP_INTERNAL_SERVER_ERROR);
+            return;
+        }
+    }
+
+    if (n > 0) {
+        return;
+    }
+
+    if (n == -1) {
+        if (err == NGX_EAGAIN) {
+            return;
+        }
+
+        ev->error = 1;
+
+    } else { /* n == 0 */
+        err = 0;
+    }
+
+    ev->eof = 1;
+    c->error = 1;
+
+    if (u->peer.connection) {
+        ngx_log_error(NGX_LOG_INFO, ev->log, err,
+                      "client prematurely closed connection, "
+                      "so upstream connection is closed too");
+        ngx_http_proxy_connect_finalize_request(r, u,
+                                           NGX_HTTP_CLIENT_CLOSED_REQUEST);
+        return;
+    }
+
+    ngx_log_error(NGX_LOG_INFO, ev->log, err,
+                  "client prematurely closed connection");
+
+    if (u->peer.connection == NULL) {
+        ngx_http_proxy_connect_finalize_request(r, u,
+                                           NGX_HTTP_CLIENT_CLOSED_REQUEST);
+    }
+}
+
+
+static void
+ngx_http_proxy_connect_rd_check_broken_connection(ngx_http_request_t *r)
+{
+    ngx_http_proxy_connect_check_broken_connection(r, r->connection->read);
+}
+
+
+static void
+ngx_http_proxy_connect_wr_check_broken_connection(ngx_http_request_t *r)
+{
+    ngx_http_proxy_connect_check_broken_connection(r, r->connection->write);
+}
+
+
+static ngx_int_t
+ngx_http_proxy_connect_handler(ngx_http_request_t *r)
+{
+    in_port_t                          (*ports)[2];
+    ngx_url_t                            url;
+    ngx_int_t                            rc;
+    ngx_uint_t                           i, allow;
+    ngx_resolver_ctx_t                  *rctx, temp;
+    ngx_http_core_loc_conf_t            *clcf;
+    ngx_http_proxy_connect_ctx_t        *ctx;
+    ngx_http_proxy_connect_upstream_t   *u;
+    ngx_http_proxy_connect_loc_conf_t   *plcf;
+
+    plcf = ngx_http_get_module_loc_conf(r, ngx_http_proxy_connect_module);
+
+    allow = 0;
+
```

Comment:
If the request does not match the `location` configuration, the content handler `ngx_http_proxy_connect_handler` of the server block will be used by default. So if the request method is not `CONNECT` and may be caused segment.

for example:

```
 server {
     listen                         3124;
     # dns resolver used by forward proxying
     #resolver                       8.8.8.8;
     #footer xxx;
     # forward proxy for CONNECT request
     proxy_connect;
     proxy_connect_allow            all;
}
```
`curl http://github.com/ -x localhost:3124`   will be  caused segment.

Checks should be added to the `ngx_http_proxy_connect_handler` entry, for example :
```
if (r->method != NGX_HTTP_CONNECT) {
    return NGX_DECLINED;
}
```

Reply:
has been fixed